src/primitives/gf32.rs

```diff
@@ -0,0 +1,481 @@
+    /// Numeric value maps to bech32 character: 31 == "l".
+    pub const L: Fe32 = Fe32(31);
+
+    /// Iterator over all field elements, in alphabetical order.
+    pub fn iter_alpha() -> impl Iterator<Item = Fe32> {
+        [
+            Fe32::A,
+            Fe32::C,
+            Fe32::D,
+            Fe32::E,
+            Fe32::F,
+            Fe32::G,
+            Fe32::H,
+            Fe32::J,
+            Fe32::K,
+            Fe32::L,
+            Fe32::M,
+            Fe32::N,
+            Fe32::P,
+            Fe32::Q,
+            Fe32::R,
+            Fe32::S,
+            Fe32::T,
+            Fe32::U,
+            Fe32::V,
+            Fe32::W,
+            Fe32::X,
+            Fe32::Y,
+            Fe32::Z,
+            Fe32::_0,
+            Fe32::_2,
+            Fe32::_3,
+            Fe32::_4,
+            Fe32::_5,
+            Fe32::_6,
+            Fe32::_7,
+            Fe32::_8,
+            Fe32::_9,
+        ]
+        .iter()
+        .copied()
+    }
+
+    /// Creates a field element from a single bech32 character.
+    pub fn from_char(c: char) -> Result<Fe32, Error> {
+        let byte = i8::try_from(u32::from(c)).map_err(|_| Error::InvalidChar(c))?;
+        let byte = byte as u8; // cast guaranteed to be ok since we started with an unsigned value
+        let u5 = u8::try_from(CHARS_INV[usize::from(byte)]).map_err(|_| Error::InvalidChar(c))?;
+        Ok(Fe32(u5))
+    }
+
+    /// Converts the field element to a lowercase bech32 character
+    pub fn to_char(self) -> char {
+        // casting and indexing fine as we have self.0 in [0, 32) as an invariant
+        CHARS_LOWER[self.0 as usize]
+    }
+
+    /// Converts the field element to a 5-bit u8, with bits representing the coefficients
+    /// of the polynomial representation.
+    pub fn to_u8(self) -> u8 { self.0 }
+
+    fn _add(self, other: Fe32) -> Fe32 { Fe32(self.0 ^ other.0) }
+
+    // Subtraction is the same as addition in a char-2 field.
+    fn _sub(self, other: Fe32) -> Fe32 { self + other }
+
+    #[cfg_attr(all(test, mutate), mutate)]
+    fn _mul(self, other: Fe32) -> Fe32 {
+        if self.0 == 0 || other.0 == 0 {
+            Fe32(0)
+        } else {
+            let log1 = LOG[self.0 as usize];
+            let log2 = LOG[other.0 as usize];
+            Fe32(LOG_INV[((log1 + log2) % 31) as usize])
+        }
+    }
+
+    #[cfg_attr(all(test, mutate), mutate)]
+    fn _div(self, other: Fe32) -> Fe32 {
+        if self.0 == 0 {
+            Fe32(0)
+        } else if other.0 == 0 {
+            panic!("Attempt to divide {} by 0 in GF32", self);
+        } else {
+            let log1 = LOG[self.0 as usize];
+            let log2 = LOG[other.0 as usize];
+            Fe32(LOG_INV[((31 + log1 - log2) % 31) as usize])
+        }
+    }
+}
+
+impl fmt::Display for Fe32 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.to_char(), f) }
+}
+
+impl str::FromStr for Fe32 {
```

Comment:
In 310012dd3e5f5725b5bde9021606ed1d5542620d:

IMO we should just drop this impl. Converting from a *string* to a single field element is a super weird thing to do, and I don't use this anywhere in my code (because I'd forgotten I"d implemented it :)).

Dropping this would also let us drop the `ExtraChar` and `InvalidLength` variants from the Error enum.